Isotonic scipy from scikit

AJTheDataGuy opened the pull request:
Please work through the following checklists. Delete anything that isn't relevant.
## Development for new xarray-based metrics
- [ ] Works with n-dimensional data and includes `reduce_dims`, `preserve_dims`, and `weights` args.
- [ ] Typehints added
- [ ] Add error handling
- [ ] Imported into the API
- [ ] Works with both `xr.DataArrays` and `xr.Datasets` if possible

## Docstrings
- [ ] Docstrings complete and follow Napoleon (google) style
- [ ] Maths equation added
- [ ] Reference to paper/webpage is in docstring. The preferred referencing style for journal articles is [APA (7th edition)](https://apastyle.apa.org/style-grammar-guidelines/references/examples/journal-article-references)
- [ ] Code example added


## Testing of new xarray-based metrics
- [ ] 100% unit test coverage
- [ ] Test that metric is compatible with dask.
- [ ] Test that metrics work with inputs that contain NaNs
- [ ] Test that broadcasting with xarray works
- [ ] Test both reduce and preserve dims arguments work
- [ ] Test that errors are raised as expected
- [ ] Test that it works with both `xr.DataArrays` and `xr.Datasets`

## Tutorial notebook 
- [ ] Short introduction to why you would use that metric and what it tells you
- [ ] A link to a reference
- [ ] A "things to try next" section at the end
- [ ] Add notebook to [Tutorial_Gallery.ipynb](https://github.com/nci/scores/blob/develop/tutorials/Tutorial_Gallery.ipynb)
- [ ] Optional - a detailed discussion of how the metric works at the end of the notebook

## Documentation
- [ ] Add the score to the [API documentation](https://github.com/nci/scores/blob/develop/docs/api.md)
- [ ] Add the score to the [included list of metrics and tools](https://github.com/nci/scores/blob/develop/docs/included.md)

Steph-Chong commented on src/scores/processing/isoreg_impl.py:
```diff
@@ -503,10 +503,13 @@ def _contiguous_mean_ir(
-    as it is about 4 times faster (since it is optimised for mean).
+    Refactored to use scipy instead of scikit learn.
+
+    IMPORTANT NOTE: the y array MUST be sorted in descending order before being passed to the function. This
+    implicitly handled by the _tidy_ir_inputs function before calling this function.
```
```suggestion
    IMPORTANT NOTE: the y array MUST be sorted in descending order before being passed to the function. This is
    implicitly handled by the _tidy_ir_inputs function before calling this function.
```
Is there meant to be an "is" between "This" and "implicitly"?

Steph-Chong commented on src/scores/processing/isoreg_impl.py:
```diff
@@ -503,10 +503,13 @@ def _contiguous_mean_ir(
+
+    IMPORTANT NOTE: the y array MUST be sorted in descending order before being passed to the function. This
+    implicitly handled by the _tidy_ir_inputs function before calling this function.
+    To make explicit could include y = np.sort(y)[::-1]
```
Re. the sentence "To make explicit could include y = np.sort(y)[::-1]", could the grammar be improved to make this clearer? (Or is it sufficiently clear as is?)

tennlee replied:
This looks to be tidied up in the latest version.

Steph-Chong reviewed (commented):
@AJTheDataGuy thanks for this PR. 

I have made two minor comments for your consideration.

tennlee commented:
To anyone reviewing this (perhaps @nicholasloveday), AJ has left some additional comments on testing in issue #636  which should be reviewed also.

nicholasloveday commented on src/scores/processing/isoreg_impl.py:
```diff
@@ -503,10 +503,13 @@ def _contiguous_mean_ir(
 ) -> np.ndarray:
     """
     Performs classical (i.e. for mean functional) contiguous quantile IR on tidied data x, y.
-    Uses sklearn implementation rather than supplying the mean solver function to `_contiguous_ir`,
```
Thanks for this update @AJTheDataGuy!
The original motivation for using the sklearn version rather than our independent implementation was the speed improvement (the scipy version didn't exist at the time). Is it possible to do a speed comparison to understand the possible impact on performance based on this change?

AJTheDataGuy commented:
Hi @nicholasloveday I just ran some tests.

Using a modified version of the testing script I ran 4 tests. All 4 tests used x,y, and weight arrays of 10 million elements each and I used cProfile to time the results for the overall script, running cProfile from the command line. The 4 tests I ran were:

1. Scikit (1 run - simulate user creating a single model )
2. Scikit (50 runs - simulate user creating 50 models)
3. Scipy (1 run - simulate user creating a single model)
4. Scipy (50 runs - simulate user creating 50 models)

1 run results:
- Scikit: 2.9s
- Scipy: 2.3s
- Winner: Scipy (20% improvement)

50 run results:
- Scikit: 105s
- Scipy: 50s
- Winner: Scipy (over 50% improvement)

These initial tests suggest Scipy is a nice performance enhance over Scikit-learn for this use case.

I've attached the python test files (start with "test") and well as the results files (start with "result") if you'd like to take a look.

[test_scipy_50.txt](https://github.com/user-attachments/files/17925979/test_scipy_50.txt)
[test_scikit_50.txt](https://github.com/user-attachments/files/17925980/test_scikit_50.txt)
[test_scikit_1.txt](https://github.com/user-attachments/files/17925981/test_scikit_1.txt)
[test_scipy_1.txt](https://github.com/user-attachments/files/17925982/test_scipy_1.txt)


[result_scikit_1.txt](https://github.com/user-attachments/files/17925987/result_scikit_1.txt)
[result_scikit_50.txt](https://github.com/user-attachments/files/17925988/result_scikit_50.txt)
[result_scipy_1.txt](https://github.com/user-attachments/files/17925989/result_scipy_1.txt)
[result_scipy_50.txt](https://github.com/user-attachments/files/17925991/result_scipy_50.txt)

@Steph-Chong  I will make those changes soon.

nicholasloveday commented:
Thanks for checking that @AJTheDataGuy - that's great that there are some performance improvements!

nicholasloveday commented on src/scores/processing/isoreg_impl.py:
```diff
@@ -503,10 +503,13 @@ def _contiguous_mean_ir(
     Performs classical (i.e. for mean functional) contiguous quantile IR on tidied data x, y.
-    Uses sklearn implementation rather than supplying the mean solver function to `_contiguous_ir`,
-    as it is about 4 times faster (since it is optimised for mean).
+    Refactored to use scipy instead of scikit learn.
```
Rather than saying that it was refactored to use scipy instead of scikit learn, perhaps it's worth saying that we use scipy instead of supplying the mean solver function to `_contiguous_ir` as it's much faster

tennlee replied:
This looks to be included in the latest version.

AJTheDataGuy commented:
Hi All - just made some updates and pushed a commit to the feature branch to my fork.

Updates:
1. Implemented the docstring comments for the _contiguous_mean_ir function as described above.
2. Fixed a merge conflict on the zenodo file
3A. I noticed that the x array parameter was no longer needed on the _contiguous_mean_ir function so removed that.
3B. Removed all fcst (x array) parameters for calls to the _contiguous_mean_ir function 
3C. In a similiar sense, the _do_ir function no longer required the fcst (x array) parameter. Removed that and changed the doc string.
3D. Removed the fcst parameter in all calls to the _do_ir function
(All changes to internal functions so hopefully the changes won't impact the end user.)
4. Updated the test for test__do_ir in the test_isoreg.py file to reflect that the fcst (x array) parameter is no longer needed.

Pytest passing all tests at the end.

Can someone please rerun the workflows?

tennlee commented:
Thanks very much for this work. It's really helpful and it's a great feeling to reduce the number of dependencies.

tennlee commented:
Fixes #636
